### PR TITLE
correctly delete fetchers

### PR DESCRIPTION
**Pull request by bridiver**
fix https://github.com/brave/brave-browser/issues/2213

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [ ] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
- enable rewards
- claim grant
- visit youtube video and make sure that is added to the table

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 1]
```diff
@@ -521,7 +521,11 @@ void RewardsServiceImpl::Shutdown() {
     }
   }
 
+  for (const auto fetcher : fetchers_) {
+    delete fetcher.first;
+  }
   fetchers_.clear();
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: is this line needed then based on the fact that you delete all fetchers in for above?

Comment (bbondy) [reply components/brave_rewards/browser/rewards_service_impl.cc]: this clears out the pointers which point to invalid memory after they were deleted.

**Comment by bbondy** [comment]
master: 3c2f5d94d7ea7b952c0b0d1055bd3c16dc194077
0.58.x: c8e55c3b264fcd3247fcb62d332ce10f9acf5305

**Comment by kjozwiak** [comment]
Approved uplift to `0.57.x` after deliberating with @rebron 👍

**Comment by bbondy** [comment]
0.57.x: ac1c4eeafca1ce249a3389f0c0a02e164bc2d4c6

**Pull request merged**